turbopack/crates/turbopack-core/src/module_graph/async_module_info.rs

```diff
@@ -72,8 +73,30 @@ async fn compute_async_module_info_single(
     // strongly-connected components, and then marking all the whole SCC as async if one of the
     // modules in the SCC is async.
 
-    let mut async_modules = self_async_modules;
     let graph_ref = graph.read();
+
+    let mut async_modules = self_async_modules;
+
+    let mut cycles = vec![];
+    let mut cycle_entries: FxHashMap<ResolvedVc<_>, SmallVec<[usize; 1]>> = Default::default();
+    graph_ref.traverse_cycles(
```

Comment:
the most efficient way to do this would be to do a traversal over the reverse graph starting at the 'self async modules'.  This would be trivially O(N) and in fact you could probably use the underlying petgraph traversals since we don't care about order, just reachability.

this shoudl be much faster

Reply:
i think https://github.com/vercel/next.js/pull/86427 achieves this same result in a single traversal